tools/cold-backup/run.sh

```diff
@@ -0,0 +1,252 @@
+    "custom-test"
+)
+
+# custom secret names
+# i.e. listener's certificate, image registry authentication
+CUSTOM_SECRETS=(
+    "external-listener-cert"
+    "registry-secret"
+    "custom-test"
+)
+
+__error() {
+    local message="$1"
+    echo "$message"
+    exit 1
+}
+
+__confirm() {
+    local ns="$1"
+    local proc="${FUNCNAME[1]}"
+    local user=$(oc whoami)
+    if [[ -n $ns && -n $user ]]; then
+        read -p "Confirm $proc of $ns namespace as $user (y/n) " reply
+        if [[ ! $reply =~ ^[Yy]$ ]]; then
+            exit 0
+        fi
+    else
+        __error "Context not ready"
+    fi
+}
+
+__export() {
+    local name="$1"
+    local label="${2-}"
+    if [[ -n $name ]]; then
+        local crs=$(oc get $name -o name -l "$label")
+        if [[ -n $crs ]]; then
+            echo "Exporting $name"
+            # delete runtime metadata expression
+            local exp="del(.metadata.namespace, .items[].metadata.namespace, \
+                .metadata.resourceVersion, .items[].metadata.resourceVersion, \
+                .metadata.selfLink, .items[].metadata.selfLink, \
+                .metadata.uid, .items[].metadata.uid, \
+                .status, .items[].status)"
+            local id=$(printf $name | sed 's/\//-/g;s/ //g')
+            oc get $crs -o yaml | yq eval "$exp" - > $__TMP/resources/$SOURCE_NS-$id.yaml
+        fi
+    fi
+}
+
+__create_pvc() {
+    local pvc="$1"
+    local size="$2"
+    local class="${3-}"
+    if [[ -n $pvc && -n $size ]]; then
+        local exp="s/\$pvc/$pvc/g; s/\$size/$size/g; /storageClassName/d"
+        if [[ -n $class ]]; then
+            exp="s/\$pvc/$pvc/g; s/\$size/$size/g; s/\$class/$class/g"
+        fi
+        echo "Creating pvc $pvc of size $size"
+        sed "$exp" $__HOME/pvc.yaml | oc create -f -
+    else
+        __error "Missing required parameters"
+    fi
+}
+
+__rsync() {
+    local pvc="$1"
+    local from="$2"
+    local to="$3"
+    if [[ -n $pvc && -n $from && -n $to ]]; then
+        local patch=$(sed "s/\$pvc/$pvc/g" $__HOME/patch.json)
+        oc run $__MPOD --image="dummy" --restart="Never" --overrides="$patch"
+        oc wait --for condition="ready" pod $__MPOD --timeout="300s"
+        oc rsync --no-perms --delete --progress $from $to
+        oc delete pod $__MPOD
+    else
+        __error "Missing required parameters"
+    fi
+}
+
+__compress() {
+    if [[ -n $BACKUP_NAME && -n $BACKUP_HOME ]]; then
+        echo "Compressing"
+        local current_dir=$(pwd)
+        cd $__TMP
+        zip -qr $BACKUP_NAME.zip *
+        mv $BACKUP_NAME.zip $BACKUP_HOME
+        cd $current_dir
+    else
+        __error "Missing required parameters"
+    fi
+}
+
+__uncompress() {
+    if [[ -n $BACKUP_NAME && -n $BACKUP_HOME ]]; then
+        echo "Uncompressing"
+        rm -rf $__TMP
+        unzip -qo $BACKUP_HOME/$BACKUP_NAME.zip -d $__TMP
+    else
+        __error "Missing required parameters"
+    fi
+}
+
+backup() {
+    __confirm $SOURCE_NS
```

Comment:
We should probably make it clear, by printing a message, that backing up will cause downtime. Otherwise you're not really getting _informed_ consent.

Reply:
Agree. Thanks.